### PR TITLE
[`MLA`] Add native FA4 support

### PR DESCRIPTION
**Pull request by vasqu**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48880) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=48880)

As per title! This can be further extended to the indexer based variation of MLA (DSA) in #48909 (see [here](https://github.com/huggingface/transformers/pull/48909/changes/e039b84f65790f47cab15258da559525cb50dd3f))

It works for everything including all the varlen ops. However, the supported dimensions and hw are quite restrictive
- Only one set of dims
- Only blackwell

### CODE REVIEW

File: src/transformers/integrations/flash_attention.py [hunk 1]
```diff
@@ -34,7 +34,11 @@ def flash_attention_forward(
     sliding_window: int | None = None,
     softcap: float | None = None,
     is_causal: bool | None = None,
-    s_aux: torch.Tensor | None = None,  # alias: learnable attention sink
+    s_aux: torch.Tensor | None = None,  # alias: learnable attention sin
```
Comment (vasqu) [inline src/transformers/integrations/flash_attention.py]:
```suggestion
    s_aux: torch.Tensor | None = None,  # alias: learnable attention sink
```

File: src/transformers/models/deepseek_v3/modeling_deepseek_v3.py [hunk 2]
```diff
@@ -470,9 +487,15 @@ def forward(
         if past_key_values is not None:
             kv_nope, k_rot = past_key_values.update(kv_nope, k_rot, self.layer_idx)
 
-        query_states = torch.cat((q_pass, q_rot), dim=-1)
-
-        key_states, value_states = self.expand_kv(kv_nope, k_rot)
+        # Based on whether we absorb the latents directly, we either expand (non FA) or prepare all necessary latents (FA)
+        query_latent_states = q_pass
+        if is_flash_attention_requested(self.config, version=4):
```
Comment (ArthurZucker) [inline src/transformers/models/deepseek_v3/modeling_deepseek_v3.py]: 🫠 
I'd love if we had 0 traces of this anywhere 
(i know its a draft!)

Comment (vasqu) [reply src/transformers/models/deepseek_v3/modeling_deepseek_v3.py]: Tried something :D maybe you like that design more

File: src/transformers/integrations/flash_attention.py [hunk 3]
```diff
@@ -34,7 +36,8 @@ def flash_attention_forward(
...
```
Comment (vasqu) [inline src/transformers/integrations/flash_attention.py]:
```suggestion
    s_aux: torch.Tensor | None = None,  # alias: learnable attention sink
```

File: tests/models/deepseek_v3/test_modeling_deepseek_v3.py [hunk 4]
```diff
@@ -69,9 +69,9 @@ def __init__(
         n_shared_experts=1,
         n_routed_experts=8,
         routed_scaling_factor=2.5,
-        kv_lora_rank=16,
+        kv_lora_rank=512,
```
Comment (vasqu) [inline tests/models/deepseek_v3/test_modeling_deepseek_v3.py:72]: only temporary so i could also use our tests as sanity check

**Review by ArthurZucker — COMMENTED** [review]
nice with is_mla! could even be a layer_type thing no?

**Comment by vasqu** [comment]
> could even be a layer_type thing no?

My concern is layer type ==> register masking, caching etc again so not sure if its really worth the overhead
